### PR TITLE
Add support for continuously starting load jobs as slots free up in the loader

### PR DESCRIPTION
**Pull request by sh-rp**
### Description
In the current implementation we more or less always start n (=max workers) load jobs, let them complete and then rerun the whole loader to schedule the next n jobs. In this PR we submit new load jobs as slots free up.

What is happening here:
* The loader now periodically checks all jobs wether they are done and schedules new ones as needed.
* Runnable Jobs now manage their own internal state (file moving still needs to be done by the loader on the main thread) and have a dedicated "run" method which is the one called on a thread.
* Renaming of the Job base classes to make clearer what is going on:
  * RunnableLoadJob: Jobs that actually do something and should be executed on a thread
  * FollowupJob: Class that creates a new job persisted to the disc
  * HasFollowupJobs (ex-FollowupJob): Trait that tells loader to look for followup Jobs
  * FinalizedJob: Not runnable because it already has an actionable state (completed, failed, retry). Used for indicating failed restored jobs, completed restored jobs and cases where nothing needs to be done
* FollowupJobs always go to the "new_jobs" folder and need to be picked up by the loader as any other jobs do, there were some just directly executed in the main thread: not good! :) We now assume that creation of FollowupJobs does not yield exceptions, I don't think this is needed and it makes the code simpler.
 * (Hopefully) simplification of the load class
 * Restoring jobs is now handled the same way as creating jobs is, this simplifies the code in a few places. Jobs that are found in the running folder now are simply started again, the bigquery job can figure out wether it needs to be resumed on its own.

Possible FollowupWork:
* Performance Tuning in loader, a lot of time is spent on parsing filenames (I think).

### CODE REVIEW

File: dlt/load/load.py
Comment (sh-rp) [inline dlt/load/load.py:400]: I am not quite sure when this needs to be run and how often, so any help there would be apreciated

File: dlt/load/load.py
Comment (sh-rp) [inline dlt/load/load.py:495]: is this the correct "package completion" condition? I think so, but am not 100% sure.

File: tests/load/test_dummy_client.py
Comment (sh-rp) [inline tests/load/test_dummy_client.py:96]: I needed to change a bunch of tests, since we do not rely on multiple executions of the run method anymore. All the changes make sense, it might be good to add a few more tests cases specific to the new implementation.

File: dlt/load/load.py
Comment (sh-rp) [inline dlt/load/load.py:470]: signal support still here :)

File: dlt/load/load.py
Comment (sh-rp) [inline dlt/load/load.py:293]: I need to collect the exception to raise in the main loop  here now, we could alternatively collect all problems we find and print them out, not just raise on exception.

**Comment by sh-rp** [comment]
Note: PR on hold until naming conventions PR is merged.

File: tests/pipeline/test_pipeline_trace.py
Comment (sh-rp) [inline tests/pipeline/test_pipeline_trace.py:371]: I am not sure why this is not working anymore, can you point me to the place in the code where these sentry events should be raised?

File: tests/pipeline/test_pipeline.py
Comment (sh-rp) [inline tests/pipeline/test_pipeline.py:1664]: in the new loader implementation, jobs that retry are moved into the new jobs folder again, I'm wondering wether "has_pending_data" should also return true if there are jobs that are marked as "to be retried"

File: tests/load/test_job_client.py
Comment (sh-rp) [inline tests/load/test_job_client.py:726]: NOTE: todo

File: tests/cli/test_pipeline_command.py
Comment (sh-rp) [inline tests/cli/test_pipeline_command.py:215]: see comment about this below

File: tests/load/test_job_client.py
Comment (sh-rp) [reply tests/load/test_job_client.py:726]: for bigquery restoring is tested now, the other just re-run the job when found in the running folder (which as discussed is the right way to do it)

File: dlt/load/utils.py
Comment (sh-rp) [inline dlt/load/utils.py:262]: Possible todo

File: dlt/load/load.py
Comment (rudolfix) [inline dlt/load/load.py:441]: this is really costly operation... make sure we update counters only when job changed state

File: dlt/load/load.py
Comment (rudolfix) [inline dlt/load/load.py:524]: I think you should really create a mock package with 50k jobs. this currently will call list new jobs 10x / second

File: dlt/load/load.py
Comment (rudolfix) [reply dlt/load/load.py:540]: this was the previous condition:
```
        if file_count == 0:
            logger.info(f"No new jobs found in {load_id}")
            return 0, []
```
so checking just new jobs. I do not fully get why you do it here again? the loop above should exit only when all jobs are completed, no?

File: dlt/load/utils.py
Comment (rudolfix) [inline dlt/load/utils.py:253]: this must be checked before you call this function. do not pass `running_jobs` here. you are mixing two separate things

**Review by rudolfix — CHANGES_REQUESTED** [review]
amazing job! and I think code got greatly simplified. also I think changes are quite comprehensible.

I reviewed load.py and I think you can take a look already

File: dlt/load/load.py
Comment (sh-rp) [reply dlt/load/load.py:209]: Yep, done!

File: dlt/destinations/sql_jobs.py
Comment (sh-rp) [reply dlt/destinations/sql_jobs.py:68]: In my view, failed jobs should be the result of something not right in the destination (=async errors). Creating a FollowUpJob should never fail, and if it does it is a programming error on our side and should create a proper exception with stacktrace and not a failed job. Followupjobs, at least for now, have no destination dependent or async component. The one exception is a full disk that might occur. If you want me to keep it though, I will put it back in.

File: dlt/load/load.py
Comment (sh-rp) [reply dlt/load/load.py:368]: I would say the old variant was not easier to test, the importing of a job was a nested function within the code that creates them. How would you split this? The part that imports the jobs only calls a function on the normalized_packages of the load_storage. If you have a good idea I'll change it.

Comment (sh-rp) [reply dlt/load/load.py:368]: That said, I will check the existing tests for the followupjobs and maybe this will result in a refactor here, let's see.

File: dlt/load/load.py
Comment (sh-rp) [reply dlt/load/load.py:450]: this should be done correctly now, I am also updating it once after the loop exits, I hope this is good

File: dlt/load/utils.py
Comment (sh-rp) [reply dlt/load/utils.py]: done

File: dlt/load/load.py
Comment (sh-rp) [reply dlt/load/load.py:256]: I have changed it this way. We could cache the list of files, let me see what a test with 30k files does.

File: tests/pipeline/test_pipeline.py
Comment (sh-rp) [reply tests/pipeline/test_pipeline.py:1768]: @rudolfix I need your opinion here.

File: tests/cli/test_pipeline_command.py
Comment (sh-rp) [reply tests/cli/test_pipeline_command.py:215]: above actually

File: dlt/common/destination/reference.py
Comment (sh-rp) [inline dlt/common/destination/reference.py:320]: client here can possibly be removed.

File: tests/load/test_dummy_client.py
Comment (sh-rp) [inline tests/load/test_dummy_client.py:124]: I did the same test on devel and it is way slower there, I'd need to dig further into it why that is the case, maybe the long sleep and the restarting of the load, or maybe something else idk.

File: dlt/common/destination/reference.py
Comment (rudolfix) [inline dlt/common/destination/reference.py:454]: this is very misleading name. it does not get any jobs. it creates new jobs. please rename

File: dlt/common/destination/reference.py
Comment (rudolfix) [inline dlt/common/destination/reference.py:457]: it does not start anything

File: dlt/destinations/impl/bigquery/bigquery.py
Comment (rudolfix) [inline dlt/destinations/impl/bigquery/bigquery.py:263]: this is 100% not needed and no calls to big query should be made from here. AFAIK you create jobs before sending them to threads and without calling `with` client so the connection in `sql_client` is closed

File: tests/load/bigquery/test_bigquery_client.py
Comment (rudolfix) [inline tests/load/bigquery/test_bigquery_client.py:292]: this test is not testing any resumes because `get_load_job` does not create anything in bigquery. you'll need to create a job that in big query, make sure it exists and then complete it. how to do that exactly IDK.

it would be cool to test it inside `load.py`. so you load a job that fails once you create another load instance run it again and make sure it is completed with retry count == 1 in file name

**Review by rudolfix — CHANGES_REQUESTED** [review]
OK there are a lot of comments. we'll need another review... overall looking good!

are you able to add a test (ie for INSERT VALUES client) that simulates killing loader when inserts are done and then resuming? can run on duckdb

File: dlt/destinations/impl/bigquery/bigquery.py
Comment (sh-rp) [reply dlt/destinations/impl/bigquery/bigquery.py:150]: done

File: dlt/destinations/impl/bigquery/bigquery.py
Comment (sh-rp) [reply dlt/destinations/impl/bigquery/bigquery.py:162]: I have added the warning and also the reason to the exception string. Imho it was right except for the missing log message.

File: dlt/destinations/impl/bigquery/bigquery.py
Comment (sh-rp) [reply dlt/destinations/impl/bigquery/bigquery.py:138]: done

File: dlt/destinations/impl/bigquery/bigquery.py
Comment (sh-rp) [reply dlt/destinations/impl/bigquery/bigquery.py:168]: I have removed it, i don't remember why this is here

File: dlt/destinations/impl/filesystem/filesystem.py
Comment (sh-rp) [reply dlt/destinations/impl/filesystem/filesystem.py:104]: The behavior of the delta load has not changed here, i think delta manages the folders for the tables.

File: dlt/destinations/impl/filesystem/filesystem.py
Comment (sh-rp) [reply dlt/destinations/impl/filesystem/filesystem.py:311]: I also don't like this. The problem is the behavior of "is_staging_destination_job" in load.py. If I add "reference" to the supported file formats for the filesystem, then reference files will also be sent to the filesystem when it is configured as staging destination in combination with another destination, although it should actually be consumed by the main destination.

File: dlt/load/load.py
Comment (sh-rp) [reply dlt/load/load.py:207]: We can put it in the init also. There will just be tons of repetitive code where each destination has to provide these vars to every load job and every load job has to pass these to "super", so I chose this way. Your call.

File: dlt/common/destination/reference.py
Comment (sh-rp) [reply dlt/common/destination/reference.py:280]: yes, as you saw below I think this should be good.

File: dlt/load/load.py
Comment (sh-rp) [reply dlt/load/load.py:210]: I looked at the query tagging stuff, I am not really sure how that is related to these run_vars? Do you want the query tags to be set this way on the job too? Or what do you have in mind?

File: dlt/load/load.py
Comment (sh-rp) [reply dlt/load/load.py:380]: In my view they are all there. I will add one new test to see what happens if creating a followupjob throws an exception and wether the package will load once the problem is resolved.

File: dlt/load/load.py
Comment (sh-rp) [reply dlt/load/load.py:552]: done

File: tests/cli/test_pipeline_command.py
Comment (sh-rp) [reply tests/cli/test_pipeline_command.py:214]: the comment right above. `def is_package_partially_loaded` returns False if all jobs in a normalized package are in the started folder. So if I have a test that needs `is_package_partially_loaded` to return true, I have to move one job into the running or completed folder. Before my changes, the job that needed retrying remained in the started_jobs folder so this was not needed. There are two places where I test this. My question above is wether we should change is_package_partially_loaded to return true also if there was at least one failed retry.

File: tests/load/bigquery/test_bigquery_client.py
Comment (sh-rp) [reply tests/load/bigquery/test_bigquery_client.py:292]: I call "r_job.run_managed(client)" here which actually connects to bigquery and resumes the job (as indicated by the _resumed_job boolean var). The test is as good as the tests that where there for bigquery before. I am happy to improve them, but I don't really want to bloat the scope of these changes with things are are not directly related, at least in my view.

File: dlt/destinations/impl/bigquery/bigquery.py
Comment (sh-rp) [reply dlt/destinations/impl/bigquery/bigquery.py]: yes that is correct, now that you mention it I have removed it. I think the test coverage for the streaming insert is not super good. we can improve that in the future maybe too, but generally speaking it is protected by the custom destination state mechanism, so I think it should be good enough for now.

File: tests/pipeline/test_pipeline_trace.py
Comment (rudolfix) [reply tests/pipeline/test_pipeline_trace.py:371]: what this test was doing: failing two jobs. those generated two log messages on error level. those messages were sent to sentry.

you changed the code so we do not log anything when messages are failed or dummy was changed in a way to not fail the packages when `FAIL_PROB` is set.

this test should work again

**Review by rudolfix — CHANGES_REQUESTED** [review]
> What exactly is the desired behavior from a conceptual standpoint when creating a followup job fails? In the old version, loading just continues and those jobs are marked as failed. I don't think this makes sense, because the load will be useless if for example a mergejob can't be created and executed. So either we decide
this is a transient problem (e.g. the user can manually fix the schema and restart he load) , in that case we just raise an exception on the main thread, so that the job that triggered the scheduling of a followupjob remains in "started_jobs" and will be rerun on the pipeline execution including the scheduling of the followupjob (it is implemented and tested like this now)
or this is a terminal problem, in which case we should also stop the load but mark the loadpackage as failed

my take:
make it a transient error

and we need how we deal with failed packages. right now we continue load and do not raise exception at the end. IMO we should change that. we should continue load but raise exception at the end automatically.

I'll write a ticket for that - it ia a breaking change

File: tests/load/test_dummy_client.py
Comment (sh-rp) [reply tests/load/test_dummy_client.py:124]: I can go back to 1 second, no problem (and maybe modify it for the speed test). If you look at "start_new_jobs", I made a change in the last round so that it checks how many slots are available before actually querying any files on the disc, so if all slots are full, it exits early and nothing is done. I'd say you can tell that it is not starving in my speed test where many files are present on the disc.

Comment (sh-rp) [reply tests/load/test_dummy_client.py:131]: The number for running empty jobs in this test are:
1000 jobs = 4s
2000 jobs = 12s
10k jobs = 372s
So it does go up quite a bit, but definately not from listing all files on every loop (as it does not do this). What I would do: Merge this like it is now, it already is faster than the old implementation, and in a followup ticket with plenty tests we cache the parsed jobs in some in memory structure that gets refreshed periodically or when the new jobs in this structure fall below the count of open slots.

File: tests/pipeline/test_pipeline_trace.py
Comment (sh-rp) [reply tests/pipeline/test_pipeline_trace.py:400]: Alright, now I get it. I fixed the dummy config and I was logging 2 exceptions per failed job, which I have changed to only log the one from the loader, now it is good again.

File: dlt/load/load.py
Comment (sh-rp) [reply dlt/load/load.py:214]: I have added a generic method for a client to prepare for a job to be executed on the thread and am using this to set the querytags, this seemed to me to be the cleanest solution

File: dlt/common/storages/load_package.py
Comment (sh-rp) [inline dlt/common/storages/load_package.py:725]: the behavior is unified now between the different package states, I'd say this is correct.

**Review by rudolfix — APPROVED** [review]
LGTM!

**Pull request merged**
